### PR TITLE
feat: port libuv threadpool

### LINKED ISSUE #8: Current AsyncWorker scheduling conflict with threads used by other libraries
Currently, AsyncWorker uses the global pthread pool for scheduling its own tasks, taking threads as they become available. This works fine if AsyncWorker is the exclusive user of multithreading, but not when the addon is linked with other libraries that also rely on threads.

Here's the general problematic scenario:

 - addon `A` has `async_method` that spawns tasks onto an `AsyncWorker`
 - it links with library `B` and invokes `B.some_method()` that uses up to `N` threads itself
 - logically, user builds addon with `PTHREAD_POOL_SIZE=N+1` (`N` threads for the library + 1 for the async op in the addon itself)

Now, if user invokes `A.async_method()`, the addon `A` will take one thread from the pool for the worker, `N` threads are left, the library takes those `N` threads, finishes the work, everything's fine.

But now another user has code that invokes `async_method()` on various inputs in parallel e.g. via `Promise.all`:

```js
Promise.all([input1, input2]).map(input => A.async_method())
```

Now, the addon `A` will take 2 threads (one per each `async_method()` invocation) to create the `AsyncWorker`s, each of those workers invokes the `B.some_method()`, each of those parallel `B.some_method()` invocations tries to allocate `N` workers == `2*N` total, but the pool only has `(N+1)-2 == N-1` threads at this point! So each of them deadlocks.

User could increase the `PTHREAD_POOL_SIZE` to `2*(N+1)`, but then the same failure will happen if they happen to invoke 3 invocations of the library in parallel, and so on.

I actually observed this behaviour with deadlocks in a real-world addon I'm cross-compiling to Wasm with emnapi right now, and it took some time to figure out what's happening.

In my case I fixed this by monkey-patching the async methods to queue them one by one when compiled to Wasm, but, ideally, this should be handled at the emnapi level.

I think the only reliable solution here is for emnapi to have a separate, exclusive, thread pool, instead of taking as many threads as it wants from the global Emscripten pool. Then, it could either use the same JS logic or even some C++ implementation of work-stealing to execute max of `EMNAPI_WORKER_POOL_SIZE` in parallel, and no more. This would allow users to configure `EMNAPI_WORKER_POOL_SIZE` to the desired number of tasks, and `PTHREAD_POOL_SIZE` to `EMNAPI_WORKER_POOL_SIZE * (N + 1)` in advance, and avoid risk of deadlocks altogether.

And then, those users who know that they're not linking against any multithreaded libraries will be able to set both `PTHREAD_POOL_SIZE` and `EMNAPI_WORKER_POOL` to `(number of CPU cores)` and get max parallelisation as they can today.

### PR DESCRIPTION
**Pull request by toyobayashi**
Fixes #8

### CODE REVIEW

File: packages/emnapi/src/emnapi.c [hunk 1]
```diff
@@ -154,69 +156,108 @@ emnapi_get_emscripten_version(napi_env env,
...
-static napi_async_work _emnapi_async_work_init(
+static napi_async_work async_work_init(
   napi_env env,
   napi_async_execute_callback execute,
   napi_async_complete_callback complete,
   void* data
 ) {
   napi_async_work work = (napi_async_work)malloc(sizeof(struct napi_async_work__));
   if (work == NULL) return NULL;
+  memset(work, 0, sizeof(struct napi_async_work__));
```
Comment (RReverser) [inline packages/emnapi/src/emnapi.c:181]: Nit: you can just use calloc instead. It's both shorter and, at least on native platforms, more efficient, so it's just a good habit.

File: packages/emnapi/src/simple-async-operation.ts [hunk 2]
```diff
@@ -1,119 +1,51 @@
...
-// #else
-  return _napi_set_last_error(env, napi_status.napi_generic_failure, 0, 0)
-// #endif
-}
-
-emnapiImplement('_emnapi_queue_async_work_js', _emnapi_queue_async_work_js, ['$PThread', '$emnapiAsyncWorkerQueue', '$emnapiGetDynamicCalls'])
-emnapiImplement('napi_cancel_async_work', napi_cancel_async_work, ['$emnapiAsyncWorkerQueue', '$emnapiGetDynamicCalls', 'napi_set_last_error'])
+  },
+  _emnapi_async_send__postset:
+    'PThread.unusedWorkers.forEach(emnapiAddSendListener);' +
```
Comment (RReverser) [inline packages/emnapi/src/simple-async-operation.ts:44]: Hm, if you switched to libuv / native threadpool, why do you still need to access Emscripten's internals here? (I can probably figure out, but quicker to ask.)

File: packages/emnapi/src/simple-async-operation.ts [hunk 3]
```diff
@@ -1,119 +1,51 @@
...
+          data
         }
-      }
-    }
-    if (ENVIRONMENT_IS_NODE) {
-      worker.on('message', worker._emnapiAsyncWorkListener)
+      })
     } else {
-      worker.addEventListener('message', worker._emnapiAsyncWorkListener, false)
+      setTimeout(() => {
```
Comment (RReverser) [inline packages/emnapi/src/simple-async-operation.ts:38]: I saw this in earlier version of the code, but forgot to ask - why are you delaying execution via `setTimeout`? `setTimeout` is guaranteed to add at least 4ms delay per invocation in modern browsers, which can be a lot for high-performance code.

If you don't need the delay, maybe you can just execute the function immediately, but if you do need it, then maybe at least use `Promise.resolve().then(() => ...)` instead which will schedule a microtask instead of a full timeout task.

File: packages/emnapi/src/simple-async-operation.ts (hunk 2 above)
Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:44]: `_emnapi_async_send` emulates `uv_async_send`. Here still need to use JavaScript to send events to the main thread, or perform an asynchronous operation in the event loop. Is there any better choice to do `uv_async_send` with emscripten API on C side?

File: packages/emnapi/src/simple-async-operation.ts (hunk 3 above)
Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:38]: _emnapi_async_send emulates uv_async_send. In my understand, `uv_async_send` actually queue asynchronous callback on event loop, not a microtask based on `Promise`, not sure whether I understand correctly.

Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:38]: > why are you delaying execution

I find the official docs said: 

http://docs.libuv.org/en/v1.x/guide/threads.html?highlight=uv_async_send#inter-thread-communication

> Remember: `uv_async_send` is also non-blocking and will return immediately.

So in order to make the JavaScript implementation behavior closer to native Node.js, I think the callback function needs to be called with delay.

Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:45]: I look into libuv source code, [`uv__async_spin`](https://github.com/libuv/libuv/blob/v1.43.0/src/unix/async.c#L84) is called both before [calling the async callback](https://github.com/libuv/libuv/blob/v1.43.0/src/unix/async.c#L157-L163) and in [`uv__async_close`](https://github.com/libuv/libuv/blob/v1.43.0/src/unix/async.c#L115-L119), is it more appropriate to use `Promise` than `setTimeout`?

Comment (RReverser) [reply packages/emnapi/src/simple-async-operation.ts:45]: I _think_ `Promise` aka microtask would be enough but I'm not sure tbh. At least, if you want a full task, you can use a `MessagePort` trick instead - it was commonly used as a polyfill for experimental `setImmediate` API.

Basically you'd just need to do:

```js
function setImmediate(fn) {
  let { port1, port2 } = new MessageChannel();
  port1.onmessage = fn;
  port2.postMessage(null);
}
```

This waits for a full event loop spin, but would be still much faster than `setTimeout`:

image

File: packages/emnapi/src/simple-async-operation.ts (hunk 2 above)
Comment (RReverser) [reply packages/emnapi/src/simple-async-operation.ts:52]: I'm not very familiar with libuv tbh. What does it do here - does it also use custom JS for sending those via V8?

Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:52]: > does it also use custom JS for sending those via V8?

It use eventfd(linux) / IOCP(windows) to wake up main thread uv loop to call the complete callback, not use v8.

Currently emnapi's async work implementation is almost a copy of Node.js, but it's difficult for me to port libuv's `async.c` which includes the thread pool async dispatching functions, so I think it's ok to leave this part be JavaScript.

Comment (RReverser) [reply packages/emnapi/src/simple-async-operation.ts:52]: Hm, is it always for sending work to the main thread or to any? If just to main, [`MAIN_THREAD_ASYNC_EM_ASM`](https://emscripten.org/docs/api_reference/emscripten.h.html?highlight=main_thread_em_asm#c.MAIN_THREAD_ASYNC_EM_ASM) could be a better option.

Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:52]: `MAIN_THREAD_ASYNC_EM_ASM` is not the expect one, it should always non-blocking.

> // If the calling thread is the main browser thread, then the specified JavaScript code is executed
> // synchronously.

`emscripten_dispatch_to_thread_async` internal also use `setTimeout`, right?

https://github.com/emscripten-core/emscripten/blob/main/system/lib/pthread/library_pthread.c#L638

and internally using the `do_dispatch_to_thread` also may call the code synchronously.

https://github.com/emscripten-core/emscripten/blob/main/system/lib/pthread/library_pthread.c#L383

Comment (RReverser) [reply packages/emnapi/src/simple-async-operation.ts:52]: > is not the expect one, it should always non-blocking.

Not sure what you mean. I man, it _is_ non-blocking, unless you run it on the same thread, but why would you use it on the same thread anyway?

Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:52]: because napi_call_threadsafe_function also can call uv_async_send, in the main thread

Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:52]: https://github.com/toyobayashi/emnapi/blob/feat-threadpool/packages/emnapi/src/emnapi.c#L631

https://github.com/toyobayashi/emnapi/blob/feat-threadpool/packages/emnapi/src/emnapi.c#L604

Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:52]: So, I mean

```js
function _emnapi_async_send (callback, data) {
  if (ENVIRONMENT_IS_PTHREAD) {
    // not main thread
    postMessage(...)
  } else {
    // main thread should delay by setImmediate or Promise
    setImmediate(() => {
      call_vp(callback, data)
    })
  }
}
```

Comment (RReverser) [reply packages/emnapi/src/simple-async-operation.ts:52]: Another thing you can do for now to keep it simple is just 

```js
MAIN_THREAD_ASYNC_EM_ASM({
  setImmediate(() => { ... });
});
```

which is not perfect perf-wise, but it would be simple and will ensure that the function is always invoked asynchronously.

Comment (toyobayashi) [reply packages/emnapi/src/simple-async-operation.ts:52]: Thank you for your patient explanation. I will take a look tomorrow after work, it's midnight in my time zone now

Comment (RReverser) [reply packages/emnapi/src/simple-async-operation.ts:52]: No worries at all, thanks!

File: packages/emnapi/src/emnapi.c [hunk 4]
```diff
@@ -154,69 +157,163 @@ emnapi_get_emscripten_version(napi_env env,
...
+  //   emscripten_dispatch_to_thread_async(main_thread,
+  //                                       __EMNAPI_ASYNC_SEND_CALLBACK_SIG,
+  //                                       callback,
+  //                                       NULL,
+  //                                       data);
+  //   // or
+  //   // MAIN_THREAD_ASYNC_EM_ASM({
+  //   //   emnapiGetDynamicCalls.call_vp($0, $1);
+  //   // }, callback, data);
+  // }
```
Comment (toyobayashi) [inline packages/emnapi/src/emnapi.c:210]: @RReverser  I left comments here. If you prefer that emnapi should not depend on `PThread` object, I would appreciate it if you could open another PR to provide help. Sorry for my broken English, forget it if there is some strange or offensive expression.

Comment (toyobayashi) [reply packages/emnapi/src/emnapi.c:210]: @RReverser [c3a3cdb1](https://github.com/toyobayashi/emnapi/commit/c3a3cdb12c79fa4881758aa75bf7790bca166ab7) At last I found `emscripten_proxy_async` and `em_proxying_queue_create` works well, but using `emscripten_proxy_get_system_queue()` will also stuck if call printf in child thread. I could not figure out why.

Comment (RReverser) [reply packages/emnapi/src/emnapi.c:210]: > If you prefer that emnapi should not depend on PThread object, I would appreciate it if you could open another PR to provide help

Heh unfortunately I'm finding a bit hard to follow how threading works in this library, so not sure I can submit a PR for that yet.

Comment (RReverser) [reply packages/emnapi/src/emnapi.c:210]: FWIW one other reason I want it to have its own pool of workers is that upstream there are talks of making it possible to use pthreads without a pthread pool, and I now have a working implementation of that at least for Node.js.

However, it seems that emnapi won't work with that approach in the current form since it manipulates internal arrays and waits until `unusedWorkers` is not empty. But, if Emscripten will use pthreads without a pthread pool, it will be empty at the start so emnapi will wait forever.

Comment (RReverser) [reply packages/emnapi/src/emnapi.c:210]: Ah, looks like I had a slightly outdated dist version and you already removed some more usages. This is looking promising, going to give it a try with the pool-less version of Emscripten.
